test/test_serial_gridreduce.cpp

```diff
@@ -0,0 +1,191 @@
+        params[i] = new_work_buf;
+      }
+    }
+
+    // There should be a single top-level ForLoop. Find its position and check
+    // that there is only one.
+    size_t top_level_loop_pos = -1;
+    for (size_t i : c10::irange(top_level_exprs.size())) {
+      Expr* expr = top_level_exprs.at(i);
+      if (expr->isA<kir::ForLoop>()) {
+        ASSERT_EQ(top_level_loop_pos, -1);
+        top_level_loop_pos = i;
+      }
+    }
+
+    // This is a poor approximation of a traversal that would appear in a
+    // lowering pass to both set the isSerial() flag on grid reductions and
+    // insert wait/release syncs.
+    kir::Scope& scope = top_level_exprs.at(top_level_loop_pos)
+                            ->as<kir::ForLoop>()
+                            ->body()
+                            .at(0)
+                            ->as<kir::ForLoop>()
+                            ->body() // BIDy
+                            .at(0)
+                            ->as<kir::ForLoop>()
+                            ->body() // i131
+                            .exprs()
+                            .back()
+                            ->as<kir::ForLoop>()
+                            ->body() // i130
+                            .exprs()
+                            .back()
+                            ->as<kir::ForLoop>()
+                            ->body() // TIDx
+                            .at(2)
+                            ->as<kir::ForLoop>()
+                            ->body(); // BIDz
+    // Now scope holds inner scope. Replace the grid reduction there
+    auto old_grop = scope.at(3)->as<kir::GridReduction>();
+    auto new_grop = IrBuilder::create<kir::GridReduction>(
+        old_grop->getReductionOpType(),
+        old_grop->init(),
+        old_grop->out(),
+        old_grop->in(),
+        new_work_buf_alloc,
+        old_grop->sync_buffer(),
+        old_grop->entrance_index(),
+        old_grop->entrances(),
+        old_grop->isAllreduce(),
+        true);
+    new_grop = new_grop->withPredicate(old_grop->predicate())
+                   ->as<kir::GridReduction>();
+    new_grop = new_grop->withWritePredicate(old_grop->writePredicate())
+                   ->as<kir::GridReduction>();
+    scope.at(3) = new_grop;
+
+    auto sync_buf = global_allocations.at(1)->buffer();
+
+    top_level_exprs.insert(
+        top_level_exprs.end() - 1,
+        IrBuilder::create<kir::BlockSerializeWait>(
+            ParallelTypeBitmap(ParallelType::BIDz), sync_buf));
+
+    top_level_exprs.push_back(IrBuilder::create<kir::BlockSerializeRelease>(
+        ParallelTypeBitmap(ParallelType::BIDz), sync_buf));
+  });
+  fe.compileFusion(fusion);
+
+  auto input = at::randn(
+      {16384, 256}, at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0));
+  auto outputs = fe.runFusion({input});
+
+  std::cout << outputs.at(0)[0] << std::endl;
+
+  testValidate(fusion, outputs, {input}, __LINE__, __FILE__);
```

Comment:
A serial reduction is expected to have a larger discrepancy than a hierarchical reduction. The question is, can the validator take that into account?

Reply:
I don't think the validator will have enough information to know whether it's a serial or hierarchical reduction, but it could just assume serial. For now I have just changed this to the more realistic 5 element reduction segment size instead of 256, matching the split-K use case much better. We can manually still play with the exaggerated case if we like.

Reply:
Yeah, if there's a workaround, that seems fine for now, but this problem may show up again and again once we move the matmul benchmarks to the new infrastructure, which optionally does the validation. We should revisit again if necessary.